src/nrepl/logging.ts

```diff
@@ -0,0 +1,54 @@
+import * as vscode from 'vscode';
+import { getStateValue, setStateValue, removeStateValue } from '../../out/cljs-lib/cljs-lib';
+import * as nodeUtil from 'util';
+
+const NREPL_MESSAGES_CHANNEL_KEY = 'nReplMessagesChannel';
+const NREPL_MESSAGES_CHANNEL_NAME = 'nREPL Messages';
+
+function getMessageChannel(): vscode.OutputChannel {
+    return getStateValue(NREPL_MESSAGES_CHANNEL_KEY);
+}
+
+function createMessageChannel(): void {
+    const channel = vscode.window.createOutputChannel(NREPL_MESSAGES_CHANNEL_NAME);
+    setStateValue(NREPL_MESSAGES_CHANNEL_KEY, channel);
+    channel.show();
+}
+
+function deleteMessageChannel(channel: vscode.OutputChannel): void {
+    channel.hide();
+    channel.dispose();
+    removeStateValue(NREPL_MESSAGES_CHANNEL_KEY);
+}
+
+function toggleEnabled(): void {
+    const channel = getMessageChannel();
+    if (channel) {
+        deleteMessageChannel(channel);
+    } else {
+        createMessageChannel();
+    }
+}
+
+function formatNreplMessage(message: any): string {
+    return nodeUtil.inspect(message, false, 2, false);
+}
+
+function log(message: any, direction: Direction): void {
+    const channel = getMessageChannel();
+    if (channel) {
+        const formattedMessage = `${direction}\n${formatNreplMessage(message)}\n`;
+        channel.appendLine(formattedMessage);
+    }
+}
+
+const enum Direction {
+    ClientToServer = 'Client -> Server',
+    ServerToClient = 'Server -> Client'
+}
```

Comment:
I'm thinking maybe use something that is a valid map key for this? Then you can select it all, wrap it in square brackets and REPL with it. Given that the rest is in some valid format. Looks a bit like JSON, right? Maybe consider making it EDN so it is extra ready for the REPL even. (Maybe that last one is for later, though.)

Reply:
So, with JSON I'm not sure if we can make the output immediately programmatically consumable because these objects are output one by one, and would need a comma separator. We could just put a comma after each and then the developer would just have to remove the last one and place the output in an array.

Of course, EDN makes this easier since it doesn't require comma separators 🤔 . I think I'd rather either leave it like it is for now and make it EDN later, or make it EDN with this PR.

Reply:
Maybe leave it as it is for now and we'll see if we or someone ever want to work with it as data? It just looked a lot like data to me so I thought I'd share the idea. That said, how about something like so?

```suggestion
const enum Direction {
    ClientToServer = '->',
    ServerToClient = '<-'
}
```

Reply:
Okay, I'll leave it for now.

With just the arrows, I don't think it would be as clear/self-explanatory in the output what each arrow means. "Does -> mean going to the server, or going to the client?"

Cider uses `-->` and `<--`, but I'd rather not do what it does if we can be more explicit.

Reply:
I think it is harder to read as it is now. Client, server, it stops me in my thinking. I think the arrows are super clear and easy to parse. But i we can amend them a bit:

```
-> sent
{ ... }

<- received
{ ... }
```

Reply:
Ah, I see. I suppose Client and Server could be a little less clear than "Calva sent this and received this." I like your above suggestion!